### ISSUE TITLE
Hardware-less testing setup

### ISSUE THREAD

**Issue by jlashner** [slaclab/pysmurf#292]
I realize this won't work for everything, but I think it would be really nice to have some sort of procedure for offline testing, since right now it's fairly difficult for me to test code, which means it's hard for me to make even small changes. Locally I'm playing around with hooking up pysmurf with the rogue EmulatorRoot to do basic coverage tests. With this we could also start to build a test suite to test really basic pysmurf functionality.

Maybe we could do something like link the pysmurf's `stream_enable_on` variable to the `StreamDataSource.SourceEnable` variable so that pysmurf can act on fake data.

Would this be valuable to anyone other than me?

**Comment by eyyoung24** [comment]
So you want to be able to stream fake data using the `stream_enable` bit? Do you know if the data emulator works for you?

**Comment by jlashner** [comment]
Yep, I've been able to setup the emulator and have been using it a little to test pysmurf functions. Yeah, I'd like to have it so that the pysmurf stream_enable functions enable the emulator streaming, but that happens to be controlled by a different rogue variable. Is there any way to link the two? I have just been manually setting the emulator's SourceEnable bit whenever I turn streaming on/off with pysmurf.

**Comment by eyyoung24** [comment]
If pymsurf is instantiated with `offline=True`, then we can probably make stream_data_on call the emulator's SourceEnable bit. We would have to test the crap out of this solution, since it would make it possible to stream fake data when you think you're streaming real data. That obviously is good for your testing, but awful awful awful for real operations. I'd want to be convinced you really need this functionality before we try to add it in.

**Comment by jlashner** [comment]
If you think this will be a problem, it definitely doesn't have to be implemented directly in pysmurf, and I can probably do something locally in my startup scripts. Is there any documentation on how to add a listener to a rogue variable? Really all I'm trying to do is add a function so that whenever `AMCc:FpgaTopLevel:AppTop:AppCore:enableStreaming` is changed, `AMCc:StreamDataSource:SourceEnable` is set to the same value.

**Comment by aer7** [comment]
Running pysmurf on fake data is wise, but I have to label this wontfix for now.

**Comment by ruck314** [comment]
The pysmurf-side request here looks resolved on `main`:

- **PR #581** ("pre/post data emulator", merged 2021-01-19, commit `c4f6b553`) was opened explicitly against this issue (its body reads *"This PR resolves #580 and #292"*) and adds set/get accessors for the rogue `PreDataEmulator` and `PostDataEmulator` devices.
- **PR #625** (merged 2021-04-13) added the companion `set/get_predata_emulator_disable` hooks.

The hooks are present on `main` today:
- `python/pysmurf/client/base/base_class.py:136-138` — `self._predata_emulator` / `self._postdata_emulator` rogue paths.
- `python/pysmurf/client/command/smurf_command.py:6751-6920` — full pysmurf API for `enable`, `disable`, `Type`, `Amplitude`, `Offset`, `Period` on both emulators.

This gives pysmurf the ability to drive a rogue emulator for offline/coverage testing, which is the concrete technical ask in the original report. The broader "build a test suite" wish is open-ended and out of scope for a single issue. Combined with the existing `wontfix` label (added 2021-11-12), this issue can likely be closed.
